compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/codegen/CodeGenerator.java

```diff
@@ -1012,7 +1012,24 @@ public void visit(BLangBinaryExpr binaryExpr) {
             genNode(binaryExpr.lhsExpr, this.env);
             genNode(binaryExpr.rhsExpr, this.env);
             RegIndex regIndex = calcAndGetExprRegIndex(binaryExpr);
-            emit(binaryExpr.opSymbol.opcode, binaryExpr.lhsExpr.regIndex, binaryExpr.rhsExpr.regIndex, regIndex);
+            int opCode = binaryExpr.opSymbol.opcode;
+            if (opCode == InstructionCodes.INT_RANGE) {
+                RegIndex endValRegIndex = binaryExpr.rhsExpr.regIndex;
+                if (OperatorKind.HALF_OPEN_RANGE.equals(binaryExpr.opKind)) {
+                    endValRegIndex = getRegIndex(TypeTags.INT);
+                    RegIndex constOneRegIndex = getRegIndex(TypeTags.INT);
+                    emit(InstructionCodes.ICONST_1, constOneRegIndex);
+                    emit(InstructionCodes.ISUB, binaryExpr.rhsExpr.regIndex, constOneRegIndex, endValRegIndex);
+                }
+                if (binaryExpr.parent instanceof BLangForeach) {
+                    // Avoid creating an array if the range is only used in a foreach statement
+                    emit(InstructionCodes.NEW_INT_RANGE, binaryExpr.lhsExpr.regIndex, endValRegIndex, regIndex);
```

Comment:
Related to above comment. This logic doesn't handle half closed case. If we do the change in desugar, we don't need to handle it here.

Reply:
The expectation here was to modify the end expression (endValRegIndex) for Half Open cases in the if block above, and use that value for both cases.

The change was done in Desugar as suggested.